hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestListBlobProducer.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.hadoop.fs.azurebfs.services.BlobProperty;
+import org.apache.hadoop.fs.azurebfs.services.ListBlobConsumer;
+import org.apache.hadoop.fs.azurebfs.services.ListBlobProducer;
+import org.apache.hadoop.fs.azurebfs.services.ListBlobQueue;
+import org.apache.hadoop.fs.azurebfs.services.PrefixMode;
+import org.apache.hadoop.fs.azurebfs.utils.TracingContext;
+
+import static org.apache.hadoop.fs.azurebfs.constants.ConfigurationKeys.FS_AZURE_PRODUCER_QUEUE_MAX_SIZE;
+
+public class ITestListBlobProducer extends AbstractAbfsIntegrationTest {
+
+  public ITestListBlobProducer() throws Exception {
+    super();
+  }
+
+  @Override
+  public void setup() throws Exception {
+    super.setup();
+    Assume.assumeTrue(
+        getFileSystem().getAbfsStore().getAbfsConfiguration().getPrefixMode()
+            == PrefixMode.BLOB);
+  }
+
+  @Test
+  public void testProducerWaitingForConsumerLagToGoDown() throws Exception {
+    Configuration configuration = Mockito.spy(getRawConfiguration());
+    configuration.set(FS_AZURE_PRODUCER_QUEUE_MAX_SIZE, "10");
+    AzureBlobFileSystem fs = (AzureBlobFileSystem) FileSystem.newInstance(
+        configuration);
+    AbfsClient client = fs.getAbfsClient();
+    AbfsClient spiedClient = Mockito.spy(client);
+    fs.getAbfsStore().setClient(spiedClient);
+    fs.setWorkingDirectory(new Path("/"));
+    fs.mkdirs(new Path("/src"));
+    ExecutorService executor = Executors.newFixedThreadPool(5);
+    List<Future> futureList = new ArrayList<>();
+    for (int i = 0; i < 20; i++) {
+      int iter = i;
+      futureList.add(executor.submit(() -> {
+        return fs.create(new Path("/src/file" + iter));
+      }));
+    }
+    for(Future future : futureList) {
+      future.get();
+    }
+
+    AtomicInteger producedBlobs = new AtomicInteger(0);
+    AtomicInteger listBlobInvoked = new AtomicInteger(0);
+
+    final ITestListBlobProducer testObj = this;
+    final ListBlobQueue queue = new ListBlobQueue(
+        fs.getAbfsStore().getAbfsConfiguration().getProducerQueueMaxSize(),
+        1);
+
+    Mockito.doAnswer(answer -> {
+      synchronized (testObj) {
+        listBlobInvoked.incrementAndGet();
+        AbfsRestOperation op = client.getListBlobs(answer.getArgument(0),
+            answer.getArgument(1), 1, answer.getArgument(3));
+        producedBlobs.incrementAndGet();
+        if(producedBlobs.get() > 10) {
+          Assert.assertTrue(queue.availableSize() > 0);
+        }
+        return op;
+      }
+        })
+        .when(spiedClient)
+        .getListBlobs(Mockito.nullable(String.class),
+            Mockito.nullable(String.class), Mockito.nullable(Integer.class),
+            Mockito.nullable(TracingContext.class));
+
+
+    ListBlobProducer producer = new ListBlobProducer("src/", spiedClient, queue,
+        null, Mockito.mock(
+        TracingContext.class));
+    ListBlobConsumer consumer = new ListBlobConsumer(queue);
+    while (producedBlobs.get() < 10) ;
```

Comment:
BusyWaiting, can be optimized using CountDownLatch

Reply:
Good comment! Will take it.
Good for especially if src code @anujmodi2021 @sreeb-msft .